### PR TITLE
Use okhttp

### LINKED ISSUE #216: httpSender throws IllegalStateException due to the usage of defaultHttpClientsd
```127.0.0.1 - - [21/Jul/2017:16:22:37 +0000] "GET /single-span HTTP/1.1" 200 72 "-" "curl/7.51.0" 159
Exception in thread "jaeger.RemoteReporter-QueueProcessor" java.lang.IllegalStateException: Invalid use of BasicClientConnManager: connection still allocated.
Make sure to release the connection before allocating another one.
	at org.apache.http.impl.conn.BasicClientConnectionManager.getConnection(BasicClientConnectionManager.java:162)
	at org.apache.http.impl.conn.BasicClientConnectionManager$1.getConnection(BasicClientConnectionManager.java:139)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:456)
	at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:906)
	at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:805)
	at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:784)
	at com.uber.jaeger.senders.HttpSender.send(HttpSender.java:67)
	at com.uber.jaeger.senders.ThriftSender.flush(ThriftSender.java:111)
	at com.uber.jaeger.reporters.RemoteReporter$FlushCommand.execute(RemoteReporter.java:164)
	at com.uber.jaeger.reporters.RemoteReporter$QueueProcessor.run(RemoteReporter.java:183)
	at java.lang.Thread.run(Thread.java:745)
```

What happened was that, the first span is sent to remote collector through the httpSender. 
The rest are not. I guess  the root cause is due to the use of defaultHttpClient because the basicClientConnectionManager is a simple connection manager that maintains only one connection at a time. Even though it is thread safe, it ought to be used by one execution thread only. 
Once we meet the illegalStateException, the rest spans cannot be sent to the remote collector through httpSender. 

Here is what i did to pass in the httpsender
```
            final Configuration.ReporterConfiguration reportConf =
                    createJaegerReportConfiguration(conf);
            final Configuration.SamplerConfiguration samplerConf =
                    new Configuration.SamplerConfiguration("const", 1);
            final B3TextMapCodec b3Codec = new B3TextMapCodec();
            GlobalTracer.register(
                    new com.uber.jaeger.Configuration(
                            componentName,
                            samplerConf,
                            reportConf).getTracerBuilder().
                            registerExtractor(Format.Builtin.HTTP_HEADERS, b3Codec).
                            registerInjector(Format.Builtin.HTTP_HEADERS, b3Codec).build());
```
could anyone help look into this please? thanks a lot

### PR DESCRIPTION
**Pull request by vprithvi**
- HttpSender uses okhttp for sending
- okhttp dependency added, also publish a shaded jar with okhttp 3.8.1
- Support configuring a HttpSender with okhttp.HttpUrl (Allows for auth, etc)
- Add a HttpSender constructor that defaults to a maxPayloadSize of 1MB
- Add tests for HttpSender

Fixes #216

### CODE REVIEW

File: jaeger-core/build.gradle [hunk 1]
```diff
@@ -19,6 +20,18 @@ dependencies {
     signature 'org.codehaus.mojo.signature:java16:1.1@signature'
 }
 
+apply plugin: 'com.github.johnrengelman.shadow'
```
Comment (black-adder) [inline jaeger-core/build.gradle]: I don't know enough to comment, why is this needed?

File: jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.thrift.TException;
+import org.glassfish.jersey.server.ResourceConfig;
+import org.glassfish.jersey.test.JerseyTest;
+import org.junit.Test;
+
+/**
+ * This tests that HttpSender can be configured to make requests
+ * and that exceptions are handled correctly.
+ * It does not verify the contents of the requests as crossdock tests at
+ * {@link com.uber.jaeger.crossdock.JerseyServer} in jaeger-crossdock verifies it.
```
Comment (vprithvi) [inline jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java]: Updated

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 3]
```diff
@@ -44,41 +43,56 @@
...
   }
 
   @Override
-  public void send(Process process, List<com.uber.jaeger.thriftjava.Span> spans) throws TException {
+  public void send(Process process, List<Span> spans) throws TException {
     Batch batch = new Batch(process, spans);
     byte[] bytes = SERIALIZER.serialize(batch);
 
+    RequestBody body = RequestBody.create(MediaType.parse("application/x-thrift"), bytes);
+    Request request = new Request.Builder().url(collectorUrl).post(body).build();
```
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: We throw a SenderException at https://github.com/uber/jaeger-client-java/blob/fa21c9a7fac68db48039ad3a0c2fa7b296dfa24c/jaeger-core/src/main/java/com/uber/jaeger/senders/ThriftSender.java#L77

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 4]
```diff
@@ -44,41 +43,56 @@
...
-      HttpPost httpPost = new HttpPost(this.collectorUri);
-      httpPost.setEntity(new ByteArrayEntity(bytes));
-      HttpResponse response = httpClient.execute(httpPost);
-      if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
-        throw new TException("Could not send " + spans.size() + " spans, response "
-            + response.getStatusLine().getStatusCode());
+      Response response = httpClient.newCall(request).execute();
+      if (!response.isSuccessful()) {
+        String responseBody = response.body() != null ? response.body().string() : "null";
+        String exceptionMessage = String.format("Could not send %d spans, response %d: %s",
```
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: True, but only one get returned. I refactored this to make it clearer.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 5]
```diff
@@ -44,41 +43,56 @@
   private static final TProtocolFactory PROTOCOL_FACTORY = new Factory();
   private static final TSerializer SERIALIZER = new TSerializer(PROTOCOL_FACTORY);
 
-  private final HttpClient httpClient = new DefaultHttpClient();
-  private final URI collectorUri;
+  private final OkHttpClient httpClient = new OkHttpClient();
+  private final HttpUrl collectorUrl;
+  private static final int ONE_MB_IN_BYTES = 1048576;
```
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: done

File: CHANGELOG.rst [hunk 6]
```diff
@@ -1,6 +1,10 @@
 Changes by Version
 ==================
 
+0.20.1
+------
+- Use okhttp for HTTP sender (https://github.com/uber/jaeger-client-java/pull/224)
+
```
Comment (vprithvi) [inline CHANGELOG.rst]: done

File: jaeger-core/build.gradle (hunk 1 above)
Comment (vprithvi) [inline jaeger-core/build.gradle]: We use this plugin to relocate the okhttp library. This enables consumers of jaeger-core to depend on a different version of okhttp if they choose to.

File: jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    new HttpSender("misconfiguredUrl");
+  }
+
+  @Test(expected = TException.class)
+  public void serverDoesntExist() throws Exception {
+    HttpSender sender = new HttpSender("http://some-server/api/traces");
+    sender.send(new Process("robotrock"), generateSpans());
+  }
+
+  private List<Span> generateSpans() {
```
Comment (vprithvi) [inline jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java]: It could, but being final doesn't guarantee that the contents of the list can't change. We don't do assertions on the content of spans so it doesn't matter for this test. If you feel strongly about it, I'll move it over.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 8]
```diff
@@ -24,61 +24,84 @@
...
+    this.collectorUrl = url;
   }
 
   @Override
-  public void send(Process process, List<com.uber.jaeger.thriftjava.Span> spans) throws TException {
+  public void send(Process process, List<Span> spans) throws TException {
     Batch batch = new Batch(process, spans);
     byte[] bytes = SERIALIZER.serialize(batch);
 
+    RequestBody body = RequestBody.create(MediaType.parse("application/x-thrift"), bytes);
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: can `MediaType.parse("application/x-thrift")` be cached?

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 9]
```diff
@@ -24,61 +24,84 @@
...
   /**
-   * @param endpoint Jaeger REST endpoint consuming jaeger.thrift, e.g http://localhost:14268/api/traces
-   * @param maxPacketSize max packet size
+   * @param url Jaeger url accepting jaeger.thrift. This needs to include query params, etc.
+   * @param maxPayloadBytes max bytes to serialize as payload
    */
-  public HttpSender(String endpoint, int maxPacketSize) {
-    super(PROTOCOL_FACTORY, maxPacketSize);
-    this.collectorUri = constructCollectorUri(endpoint);
+  public HttpSender(HttpUrl url, int maxPayloadBytes) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: any reason for this to be public? Why even be a separate ctor from L63? If we need it we can add it in the future.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 10]
```diff
@@ -24,61 +24,84 @@
...
+    super(PROTOCOL_FACTORY, maxPayloadBytes);
+    if (url == null) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: if this ctor is not public, then nulll would never happen

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 3 above)
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: as I recall, the maxPacketSize is only relevant to the base class that decides when to send the batch vs. just append the span to the buffer. The actual request could be >1Mb.

Btw, where did the 1Mb number come from? Previous version of this class didn't seem to have such default, so why do we need to introduce it? There's no particular meaning in 1MB, it's really up to the user of this class to decide how much or little they should be batching.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 11]
```diff
@@ -24,61 +24,84 @@
...
-      httpPost.setEntity(new ByteArrayEntity(bytes));
-      HttpResponse response = httpClient.execute(httpPost);
-      if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
-        throw new TException("Could not send " + spans.size() + " spans, response "
-            + response.getStatusLine().getStatusCode());
-      }
+      response = httpClient.newCall(request).execute();
     } catch (IOException e) {
-      throw new TException("Could not send " + spans.size() + ", spans", e);
+      throw new TException("Could not make request to send " + spans.size() + ", spans", e);
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: no comma before `" spans"`

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 12]
```diff
@@ -24,61 +24,84 @@
...
 import org.apache.thrift.protocol.TBinaryProtocol.Factory;
 import org.apache.thrift.protocol.TProtocolFactory;
 
 public class HttpSender extends ThriftSender {
+
   private static final String HTTP_COLLECTOR_JAEGER_THRIFT_FORMAT_PARAM = "format=jaeger.thrift";
   private static final TProtocolFactory PROTOCOL_FACTORY = new Factory();
   private static final TSerializer SERIALIZER = new TSerializer(PROTOCOL_FACTORY);
+  private static final int ONE_MB_IN_BYTES = 1048576;
+  private final OkHttpClient httpClient = new OkHttpClient();
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: this work was prompted by a request to better handle connections. What does okhttp do by default? Maybe add a comment explaining that.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 13]
```diff
@@ -24,61 +24,84 @@
...
+import okhttp3.Response;
 import org.apache.thrift.TException;
 import org.apache.thrift.TSerializer;
 import org.apache.thrift.protocol.TBinaryProtocol.Factory;
 import org.apache.thrift.protocol.TProtocolFactory;
 
 public class HttpSender extends ThriftSender {
+
   private static final String HTTP_COLLECTOR_JAEGER_THRIFT_FORMAT_PARAM = "format=jaeger.thrift";
   private static final TProtocolFactory PROTOCOL_FACTORY = new Factory();
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: * please change to `new TBinaryProtocol.Factory();` for better readability

File: jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java (hunk 7 above)
Comment (yurishkuro) [inline jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java]: new objects for each test feel better, mem/perf are not a concern in the test

File: jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    HttpSender sender = new HttpSender(target("/api/tracesErr").getUri().toString());
+    sender.send(new Process("robotrock"), generateSpans());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void misconfiguredUrl() throws Exception {
+    new HttpSender("misconfiguredUrl");
+  }
+
+  @Test(expected = TException.class)
```
Comment (yurishkuro) [inline jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java]: aside from the test name, there's no difference in assertions from the `sendServerError` test. Maybe they should be expecting different exception types?

File: CHANGELOG.rst [hunk 15]
```diff
@@ -1,6 +1,11 @@
 Changes by Version
 ==================
 
+0.20.1
```
Comment (pavolloffay) [inline CHANGELOG.rst]: `0.20.6`

File: jaeger-core/build.gradle (hunk 1 above)
Comment (pavolloffay) [inline jaeger-core/build.gradle]: I don't think we should use shadow. Is there even request for it? 

cc @jpkrohling

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 9 above)
Comment (pavolloffay) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: I don't see a point of this constructor either.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 16]
```diff
@@ -24,61 +24,84 @@
...
 
-  private URI constructCollectorUri(String endpoint) {
-    try {
-      return new URI(String.format("%s?%s", endpoint, HTTP_COLLECTOR_JAEGER_THRIFT_FORMAT_PARAM));
-    } catch (URISyntaxException e) {
-      throw new IllegalArgumentException("Wrong collector host", e);
+    if (!response.isSuccessful()) {
+      String responseBody;
+      try {
+        responseBody = response.body() != null ? response.body().string() : "null";
```
Comment (pavolloffay) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: you can simply write this `responseBody = String.valueOf(response.body().string());`

File: CHANGELOG.rst (hunk 15 above)
Comment (pavolloffay) [inline CHANGELOG.rst]: or maybe `0.21.0` It depends of `HttpSender` constructor

File: jaeger-core/build.gradle (hunk 1 above)
Comment (vprithvi) [inline jaeger-core/build.gradle]: @pavolloffay I think we should because http libraries are extremely common and prone to conflict. 
Even in this project, without shading, the build breaks because `jaeger-zipkin` depends on `jaeger-core`, and the okhttp versions conflict (3.4.x vs 3.8.x). 
Also, the shaded artifact is published under a different classifier, so it wouldn't affect people who don't need it.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 10 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: Your statement incorrect, why do you think so?

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 8 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: Indeed it can, will update.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 11 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: fixed

File: jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java (hunk 14 above)
Comment (vprithvi) [inline jaeger-core/src/test/java/com/uber/jaeger/senders/HttpSenderTest.java]: This is because the exception message varies between them. We can assert on the messages, but I don't see a lot of benefit doing so.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 16 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: I could, but it would not protect us from `NPE`.
Specifically, `response.body()` itself can be `null`, so the `.string()` access can throw a `NPE`. Also, `body.string()` consumes the response and is different from `body.toString()`.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 9 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: This is an oversight, will refactor

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 3 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: >the maxPacketSize is only relevant to the base class that decides when to send the batch vs. just append the span to the buffer

Yes, we do the following in `append`
```
maxSpanBytes = maxPacketSize - EMIT_BATCH_OVERHEAD; 
 if (byteBufferSize <= maxSpanBytes) {
      spanBuffer.add(thriftSpan);
      if (byteBufferSize < maxSpanBytes) {
        return 0;
      }
      return flush();
    }
```

>The actual request could be >1Mb.

There might be some padding, but I expect this to be pretty close to 1Mb. Why do you say that the request could be >1Mb? 

>Btw, where did the 1Mb number come from?

It's an arbitrary number, some http servers have a limit of 2Mb on post payloads, so I thought 1Mb is a reasonable number.

>Previous version of this class didn't seem to have such default, so why do we need to introduce it?

This is incorrect, the previous version has a default set by the base class at 65K. 

>There's no particular meaning in 1MB, it's really up to the user of this class to decide how much or little they should be batching.

I'd argue that there is no particular meaning in 65K for http connections either. I'm open to other suggestions, or a justification to why it needs to default to 65K for http connections as well.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 13 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: done

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 16 above)
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: What @pavolloffay meant was `responseBody = String.valueOf(response.body());`

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 17]
```diff
@@ -24,61 +24,81 @@
...
   }
 
   @Override
-  public void send(Process process, List<com.uber.jaeger.thriftjava.Span> spans) throws TException {
+  public void send(Process process, List<Span> spans) throws TException {
     Batch batch = new Batch(process, spans);
     byte[] bytes = SERIALIZER.serialize(batch);
 
+    RequestBody body = RequestBody.create(MEDIA_TYPE_THRIFT, bytes);
+    Request request = new Request.Builder().url(collectorUrl).post(body).build();
```
Comment (jpkrohling) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: Is it possible to store `Request.Builder().url(collectorUrl)` in a static/instance field, instead of inside the method? Would this help performance? I assume the `collectorUrl` won't change that often, so, you wouldn't need to rebuild it for every `#send`.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 16 above)
Comment (jpkrohling) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: How expensive is `#body()` in OkHttp? Would it the second call be a cached version of the body, or would it re-parse the response?

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 18]
```diff
@@ -24,61 +24,81 @@
...
-   * @param maxPacketSize max packet size
+   * @param endpoint Jaeger REST endpoint consuming jaeger.thrift, e.g
+   * http://localhost:14268/api/traces
+   * @param maxPayloadBytes max bytes to serialize as payload
+   * @param client a client used to make http requests
    */
-  public HttpSender(String endpoint, int maxPacketSize) {
-    super(PROTOCOL_FACTORY, maxPacketSize);
-    this.collectorUri = constructCollectorUri(endpoint);
+  public HttpSender(String endpoint, int maxPayloadBytes, OkHttpClient client) {
```
Comment (jpkrohling) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: You are effectively breaking the compatibility with older clients. Is that intentional? Would it make sense to create a new constructor with the OkHttpClient, changing the existing one to create an empty OkHttpClient (or obtain it via a singleton provider perhaps?)

Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: +1 - passing okhttpclient to a public constructor is a functionality nobody asked for, thus is unnecessary increase in the API surface. But if the old class allowed overriding the buffer size then we should keep that.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 19]
```diff
@@ -24,61 +24,81 @@
...
+    super(PROTOCOL_FACTORY, maxPayloadBytes);
+    this.collectorUrl = HttpUrl.parse(String.format("%s?%s", endpoint, HTTP_COLLECTOR_JAEGER_THRIFT_FORMAT_PARAM));
+    if (this.collectorUrl == null) {
+      throw new IllegalArgumentException("Could not parse url.");
```
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: I didn't do so because it can be a bit misleading. `HttpUrl.parse` returns a `null` when it's unable to parse successfully.

Comment (jpkrohling) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: You can still throw a NPE (which is appropriate, as *your* code expects it to not be null), and still use a message like 'Could not parse the URL' :-)

Another option is to follow OkHttp semantics and return null at this point, without throwing an exception (not my taste, but still possible).

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 18 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: > You are effectively breaking the compatibility with older clients. Is that intentional? 

@jpkrohling  It is because I assumed nobody was using this (especially with #216). I could do what you suggest, and mark it deprecated.

>+1 - passing okhttpclient to a public constructor is a functionality nobody asked for, thus is unnecessary increase in the API surface. 

@yurishkuro This is true, I only exposed it because it allows end users to easily change how data is sent to collectors. It also allows for easy configuration of authentication, proxies, etc. 
To minimize the API surface, we could simply ask for easy of these configuration options when somebody requests for it, and then pass in parameters to okhttp builder.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 19 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: True, I'll update

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 16 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: `#body()` returns a reference to a one shot stream reader so calling it many times is cheap. The `body().string()` consumes it.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 17 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: Looking at the builder code, I don't see why this won't work. I'll update.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 19 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: Thinking more, `IllegalStateException` seems more correct.

Comment (jpkrohling) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: `IllegalStateException` would make sense if an object has been created, but suddenly had bad data in its internal data structure. You are here on the constructor, so, there's no "state" yet ;-)

Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: Hah, perhaps IllegalArgument makes sense 

From javadoc, 
> * Thrown to indicate that a method has been passed an illegal or
 * inappropriate argument.

I'd argue that an unparseable url is inappropriate, no? :)

Comment (jpkrohling) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: It's certainly better than `IllegalStateException`, yes :) I *personally* prefer to signal unexpected nulls as NPEs and use `IllegalArgumentException` for other validation problems (business constraint violations, for instance). But it's your call ;)

Comment (objectiser) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: Agree - `IllegalArgumentException` is more informative.

Comment (black-adder) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: I think `ArrayIndexOutOfBoundsException` would be most apt.

IllegalArgumentException takes my vote

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java (hunk 18 above)
Comment (vprithvi) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: I made the ctor with the client private. Let's discuss about whether to expose it in a separate issue.
